### PR TITLE
refactored ArchimatchUser CreatePW and ResetPW and added update-data 

### PR DESCRIPTION
**Pull request by ahmedhamilastage**

### CODE REVIEW

File: app/users/services/ArchimatchUserService.py [hunk 1]
```diff
@@ -72,70 +59,43 @@ def archimatch_user_create_password(cls, request):
...
-                response_data = {
-                    "message": "User does not exist",
-                    "status_code": status.HTTP_404_NOT_FOUND,
-                }
-                return Response(
-                    response_data["message"],
-                    status=response_data["status_code"],
-                )
+            if not ArchimatchUser.objects.filter(email=email).exists():
+                raise ValidationError("User does not exist")
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:72]: raise NotFound

File: app/users/services/ArchimatchUserService.py [hunk 2]
```diff
@@ -72,70 +59,43 @@ def archimatch_user_create_password(cls, request):
...
 
-            user = ArchimatchUser.objects.get(email=req_email)
-            password = data.get("password")
-            confirm_password = data.get("confirm_password")
+            user = ArchimatchUser.objects.get(email=email)
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:74]: user = ArchimatchUser.objects.get(email=email)
if not user:
                raise NotFound("User does not exist")

File: app/users/services/ArchimatchUserService.py [hunk 3]
```diff
@@ -72,70 +59,43 @@ def archimatch_user_create_password(cls, request):
...
-                    "status_code": status.HTTP_200_OK,
-                    "tokens": tokens,
-                }
+                tokens = cls.generate_tokens_for_user(email, password)
                 return Response(
-                    response_data,
-                    status=response_data["status_code"],
+                    {
+                        "message": "Password successfully updated",
+                        "tokens": tokens,
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:87]: why tokens  plural

File: app/users/services/ArchimatchUserService.py [hunk 4]
```diff
@@ -72,70 +59,43 @@ def archimatch_user_create_password(cls, request):
...
             else:
-                response_data = {
-                    "message": "Passwords do not match",
-                    "status_code": status.HTTP_400_BAD_REQUEST,
-                }
-                return Response(
-                    response_data["message"],
-                    status=response_data["status_code"],
-                )
+                raise ValidationError("Passwords do not match")
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:92]: detail=

File: app/users/services/ArchimatchUserService.py [hunk 5]
```diff
@@ -72,70 +59,43 @@ def archimatch_user_create_password(cls, request):
...
+        except ValidationError as e:
+            raise e
 
-        except APIException as e:
-            return Response(
-                {"message": str(e)},
-                status=e.status_code,
-            )
+        except Exception as e:
+            raise APIException("Error creating password", str(e))
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:98]: detail=

File: app/users/services/ArchimatchUserService.py [hunk 6]
```diff
@@ -153,71 +113,52 @@ def archimatch_user_reset_password(cls, request):
...
-                response_data = {
-                    "message": {"message": "Incorrect old password"},
-                    "status_code": status.HTTP_400_BAD_REQUEST,
-                }
-                return Response(
-                    response_data["message"],
-                    status=response_data["status_code"],
-                )
-
+                raise ValidationError("Incorrect old password")
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:127]: detail=

File: app/users/services/ArchimatchUserService.py [hunk 7]
```diff
@@ -153,71 +113,52 @@ def archimatch_user_reset_password(cls, request):
...
+                return Response(
+                    {"message": "Password successfully updated"},
+                    status=status.HTTP_200_OK,
+                )
             else:
-                response_data = {
-                    "message": {"message": "New passwords do not match"},
-                    "status_code": status.HTTP_400_BAD_REQUEST,
-                }
+                raise ValidationError("New passwords do not match")
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:137]: detail=

File: app/users/services/ArchimatchUserService.py [hunk 8]
```diff
@@ -153,71 +113,52 @@ def archimatch_user_reset_password(cls, request):
...
             return Response(
-                {"message": str(e)},
-                status=e.status_code,
+                {"message": "User data successfully updated", "user": serializer.data},
+                status=status.HTTP_200_OK,
             )
+        except ValidationError as e:
+            raise e
+        except Exception as e:
+            raise APIException("Error updating user data", str(e))
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:164]: detail=

File: app/users/services/ArchimatchUserService.py [hunk 9]
```diff
@@ -153,71 +113,52 @@ def archimatch_user_reset_password(cls, request):
...
+    @classmethod
+    def archimatch_user_update_data(cls, request):
+        """
+        Updates the basic data for a user based on provided data.
+        """
+        try:
+            user = request.user
+            data = request.data
+            if not data:
+                raise ValidationError("At least one field must be provided for update.")
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:153]: detail=

File: app/users/services/ArchimatchUserService.py [hunk 10]
```diff
@@ -153,71 +113,52 @@ def archimatch_user_reset_password(cls, request):
...
-                    "status_code": status.HTTP_400_BAD_REQUEST,
-                }
-                return Response(
-                    response_data["message"],
-                    status=response_data["status_code"],
-                )
-
+                raise ValidationError("Incorrect old password")
             # Set new password if conditions are met
             if new_password == confirm_new_password:
```
Comment (MohamedMegdich1) [inline app/users/services/ArchimatchUserService.py:129]: if new_password != confirm_new_password:
raise ValidationError("New passwords do not match")

File: app/users/serializers/ArchimatchUserPWSerializer.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        password: The password to set for the user. (write-only)
+        confirm_password: Confirmation of the password. (write-only)
+
+    Notes:
+        - Both `password` and `confirm_password` are write-only fields, meaning they are
+         used for input only.
+    """
+
+    old_password = serializers.CharField(write_only=True)
+    new_password = serializers.CharField(write_only=True)
```
Comment (MohamedMegdich1) [inline app/users/serializers/ArchimatchUserPWSerializer.py:64]: Add validation to test for password equality

**Review by MohamedMegdich1 — CHANGES_REQUESTED** [review]

File: app/users/services/ArchimatchUserService.py (hunk 2 above)
Comment (ahmedhamilastage) [reply app/users/services/ArchimatchUserService.py:74]: already checked existence : 

```
# Check if user exists
            if not ArchimatchUser.objects.filter(email=email).exists():
                raise NotFound("User does not exist")

            user = ArchimatchUser.objects.get(email=email)
```

File: app/users/services/ArchimatchUserService.py (hunk 3 above)
Comment (ahmedhamilastage) [reply app/users/services/ArchimatchUserService.py:87]: access & refresh

File: app/users/services/ArchimatchUserService.py (hunk 10 above)
Comment (ahmedhamilastage) [reply app/users/services/ArchimatchUserService.py:129]: same thing here : 

```
if new_password == confirm_new_password:
                user.set_password(new_password)
                user.save()
                return Response(
                    {"message": "Password successfully updated"},
                    status=status.HTTP_200_OK,
                )
            else:
                raise ValidationError(detail="New passwords do not match")
```

File: app/users/serializers/ArchimatchUserPWSerializer.py (hunk 11 above)
Comment (ahmedhamilastage) [reply app/users/serializers/ArchimatchUserPWSerializer.py:64]: in the service

**Review by MohamedMegdich1 — APPROVED** [review]

**Pull request merged**